### PR TITLE
Bookmarks - Early access message

### PR DESCRIPTION
**Pull request by ashiagr**
## Description

This adjusts the upsell message for early access availability as per [this grid](https://github.com/Automattic/pocket-casts-android/blob/c4ced1411b54cdf0beed22fe68c3f5fe011c7c5f/modules/features/player/src/test/java/au/com/shiftyjelly/pocketcasts/player/view/bookmark/components/UpsellViewModelTest.kt#L56-L61). 

Slack ref: p1695678047868919/1695437361.819799-slack-C055BDMU095

## Testing Instructions

Prerequisites
- Enable bookmark and patron features 
- Set `Feature.BOOKMARK_ENABLED.tier` temporarily to `FeatureTier.Plus(ReleaseVersion(7, 49))`

[Smoke test cases](https://docs.google.com/spreadsheets/d/1jdTUGVNhhJm6YnXAdUU93QtSvRuY-RW6HSb5bz8Jrd4/edit#gid=0)

#### Beta release

1. Ensure `versionName` in `version.properties` file is set to `7.49-rc-1`
2. Install the app and sign out if already signed in
3. Smoke test the app and notice `Plus` plan is shown without early access message for bookmarks
4. Login with `Plus` account
5. ✅ Notice that bookmarks feature is available for `Plus` user
6. Sign out

#### Early access production release

8. Update `versionName` to `7.49` in `version.properties` file
9. Install the app
10. Smoke test the app and notice `Patron` plan is shown with an early access message for bookmarks
11. Login with `Plus` account
12. ✅ Notice that bookmarks feature is unavailable for `Plus` user
13. Login with `Patron` account
14. ✅ Notice that bookmarks feature is available for `Patron` user
13. Sign out

#### Full access production release

15. Update `versionName` to `7.50` in `version.properties` file
16. Install the app
17. Smoke test the app and notice `Plus` plan is shown without an early access message for bookmarks
18. Login with `Plus` account
19. ✅ Notice that bookmarks feature is available for `Plus` user

## Screenshots or Screencast 

Early access production release | F
-----|-----
 | 

## Checklist
- [ ] If this is a user-facing change, I have added an entry in CHANGELOG.md
- [x] Ensure the linter passes (`./gradlew spotlessApply` to automatically apply formatting/linting)
- [x] I have considered whether it makes sense to add tests for my changes
- [x] All strings that need to be localized are in `modules/services/localization/src/main/res/values/strings.xml`
- [ ] Any jetpack compose components I added or changed are covered by compose previews
 
#### I have tested any UI changes...

- [ ] with different themes
- [ ] with a landscape orientation
- [ ] with the device set to have a large display and font size
- [ ] for accessibility with TalkBack

### CODE REVIEW

File: modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/bookmark/components/UpsellViewModel.kt [hunk 1]
```diff
@@ -19,17 +20,25 @@ import kotlinx.coroutines.flow.stateIn
...
     private val analyticsTracker: AnalyticsTrackerWrapper,
     private val subscriptionManager: SubscriptionManager,
+    private val feature: FeatureWrapper,
+    private val releaseVersion: ReleaseVersionWrapper,
 ) : ViewModel() {
 
     private val _state: MutableStateFlow<UiState> = MutableStateFlow(UiState.Loading)
     val state: StateFlow<UiState> = _state
 
+    private val earlyAccessEndDate = LocalDate.of(2023, 10, 30).atStartOfDay().atZone(ZoneId.systemDefault())
```
Comment (ashiagr) [inline modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/bookmark/components/UpsellViewModel.kt:39]: This is a placeholder date.

**Review by mchowning — CHANGES_REQUESTED** [review]
I think there are just a few issues with the logic around the early access period that we need to fix and this will be good.

I tested with a hardcoded early access release of `7.49`, and looking at builds of the app for different releases, these were my results.

| Release   | Free User      | Plus User      | Patron User  |
|----------|----------------|----------------|--------------|
| 7.48-rc-1 | Plus Upsell    | ⚠️ Plus Upsell | Available    |
| 7.48      | ⚠️ Plus Upsell | ⚠️ Available   | ⚠️ Available |
| 7.49-rc-1 | Plus Upsell    | Available      | Available    |
| 7.49      | Patron Upsell  | Patron Upsell  | Available    |
| 7.50-rc-1 | Plus Upsell    | ⚠️ Plus Upsell | Available    |
| 7.50      | Plus Upsell    | Available      | Available    |

I think that for now we can probably ignore the pre-early access period release issues and just file an issue for them if we want. It would be nice to get that working as well so we don't have to think about it anymore, but it's not an immediate concern since we're not doing multiple releases of beta testing for bookmarks.

Plus beta users getting an upsell after the early access period is the only blocker I believe. I left a comment below where I think the logic needs to be tweaked.

Otherwise, this looks good. Nice work!

File: modules/services/featureflag/src/main/java/au/com/shiftyjelly/pocketcasts/featureflag/Feature.kt [hunk 2]
```diff
@@ -63,10 +63,11 @@ enum class Feature(
                     // Patron features can only be used by Patrons
                     FeatureTier.Patron -> false
 
-                    // Plus users cannot use Plus features during early access for patrons
+                    // Plus users cannot use Plus features during early access for patrons except when the app is in beta
                     is FeatureTier.Plus ->
                         FeatureFlag.isEnabled(feature) &&
-                            !feature.tier.patronExclusiveAccessRelease.matchesCurrentReleaseForEarlyPatronAccess()
+                            (ReleaseVersion.currentReleaseVersion.releaseCandidate != null && feature.tier.patronExclusiveAccessRelease.matchesCurrentReleaseForEarlyPatronAccess()) ||
+                            (ReleaseVersion.currentReleaseVersion.releaseCandidate == null && !feature.tier.patronExclusiveAccessRelease.matchesCurrentReleaseForEarlyPatronAccess())
```
[the remark this answers is not in the document]
Comment (ashiagr) [reply modules/services/featureflag/src/main/java/au/com/shiftyjelly/pocketcasts/featureflag/Feature.kt:70]: These are great suggestions! I'll update the PR after testing different states, this is a bit confusing.

[the remark this answers is not in the document]
Comment (ashiagr) [reply modules/services/featureflag/src/main/java/au/com/shiftyjelly/pocketcasts/featureflag/Feature.kt]: I updated the PR, it is ready for another look. 🙏 

I also updated `FeatureTest` and removed `Ignored` annotation for the failing tests.

File: modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/bookmark/components/UpsellViewModel.kt [hunk 3]
```diff
@@ -25,6 +28,8 @@ import javax.inject.Inject
 class UpsellViewModel @Inject constructor(
     private val analyticsTracker: AnalyticsTrackerWrapper,
     private val subscriptionManager: SubscriptionManager,
+    private val feature: FeatureWrapper,
+    private val releaseVersion: ReleaseVersionWrapper,
 ) : ViewModel() {
```
Comment (mchowning) [inline modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/bookmark/components/UpsellViewModel.kt:33]: Instead of having the `FeatureWrapper` class with a hardcoded reference to the bookmarks feature, we could just have an alternate constructor for testing purposes:

``` kotlin
class UpsellViewModel @VisibleForTesting internal constructor(
    private val analyticsTracker: AnalyticsTrackerWrapper,
    private val subscriptionManager: SubscriptionManager,
    private val bookmarksFeature: Feature,
    private val releaseVersion: ReleaseVersionWrapper,
) : ViewModel() {

    @Inject
    constructor(
        analyticsTracker: AnalyticsTrackerWrapper,
        subscriptionManager: SubscriptionManager,
        releaseVersion: ReleaseVersionWrapper,
    ) : this(
        analyticsTracker = analyticsTracker,
        subscriptionManager = subscriptionManager,
        bookmarksFeature = Feature.BOOKMARKS_ENABLED,
        releaseVersion = releaseVersion,
    )
```

Sharing this more just as food for thought. I tend to prefer avoiding wrapper classes for testing when possible, but having alternate constructors isn't an ideal solution either since they have a lot of repetition.

The other option that comes to mind would be having a custom method in a hilt module that takes in the injected dependencies and uses them to manually construct and return the `UpsellViewModel`. That has the benefit of not bloating the `UpsellViewModel` class, but I'm not sure if adding that bit of DI-indirection might end up making this seem more confusing.

Long story short, if I were writing this I probably would have used one of the alternatives that avoided using a wrapper class, but I realize the alternatives have their own problems, so no need to change this, I'm just talking about loud here.

**Review by mchowning — APPROVED** [review]
Nice work Ashita! This is working pretty well. In my testing, all of the current and future release test cases are passing! 👍 

These tests are with a `7.49` patron early access release.

| Release   | Free User        | Plus User        | Patron User  |
| ----------- | ------------------ | ------------------ | --------------|
| 7.48-rc-1 | Plus Upsell      | Available        | Available    |
| 7.48      | ⚠️ Patron Upsell | ⚠️ Patron Upsell | ⚠️ Available |
| 7.49-rc-1 | Plus Upsell      | Available        | Available    |
| 7.49      | Patron Upsell    | Patron Upsell    | Available    |
| 7.50-rc-1 | Plus Upsell      | Available        | Available    |
| 7.50      | Plus Upsell      | Available        | Available    |

Obviously, we're still not handling the case where we want to have a feature in beta for multiple releases, but that still feels like an edge case we probably won't actually encounter. ~~I do wonder if it might be worth adding an init method to `FeatureTier.Plus` that throws an exception if the `patronExclusiveAccessRelease` parameter is a future release. That way there wouldn't be any risk of us leaking a feature early.~~ Nevermind, that's not a good idea because we'll (hopefully) always be merging these changes before we do code freeze on the release.

I left a few other comments, but they have more to do with making the code easier to follow since there is some pretty tricky logic involved. Those comments are not blockers and don't have anything to do with functionality though, so I think it would be fine to go ahead and merge this as-is and if you want to do any updates based on my comments, we can just do those in a small new PR.

File: modules/services/utils/src/main/java/au/com/shiftyjelly/pocketcasts/utils/featureflag/Feature.kt [hunk 4]
```diff
@@ -70,10 +87,20 @@ enum class Feature(
...
 
-                    // Plus users cannot use Plus features during early access for patrons
-                    is FeatureTier.Plus ->
+                    // Plus users cannot use Plus features during early access for patrons except when the app is in beta
+                    is FeatureTier.Plus -> {
+                        val isReleaseCandidate = releaseVersion.currentReleaseVersion.releaseCandidate != null
+                        val relativeToEarlyAccess = feature.tier.patronExclusiveAccessRelease?.let {
+                            releaseVersion.currentReleaseVersion.comparedToEarlyPatronAccess(it)
+                        }
                         FeatureFlag.isEnabled(feature) &&
```
[the remark this answers is not in the document]
Comment (ashiagr) [reply modules/services/utils/src/main/java/au/com/shiftyjelly/pocketcasts/utils/featureflag/Feature.kt:96]: I get what you're saying. I renamed this method from `isAvailable` to `isUserEntitled` and took out the `isEnabled` checks from this method in https://github.com/Automattic/pocket-casts-android/pull/1438/commits/e1f217d687f87bd4031f97ad4ee898affd0be1da.

File: modules/services/utils/src/main/java/au/com/shiftyjelly/pocketcasts/utils/featureflag/Feature.kt [hunk 5]
```diff
@@ -52,9 +52,26 @@ enum class Feature(
...
+            EarlyAccessState.Before,
+            EarlyAccessState.During -> !isReleaseCandidate
+            EarlyAccessState.After -> false
+        }
+    }
+
     companion object {
 
-        fun isAvailable(feature: Feature, userTier: UserTier) = when (userTier) {
+        fun isAvailable(
```
[the remark this answers is not in the document]
Comment (ashiagr) [reply modules/services/utils/src/main/java/au/com/shiftyjelly/pocketcasts/utils/featureflag/Feature.kt]: >  I feel like having both this Feature::isAvailable method and the FeatureFlag::isEnabled methods is a bit confusing since they sound so similar. What would you think about renaming this method to something like isUserEntitled

Same as above,  I renamed this method from `isAvailable` to `isUserEntitled` in https://github.com/Automattic/pocket-casts-android/commit/e1f217d687f87bd4031f97ad4ee898affd0be1da. 

> Also, what would you think about moving this method to the FeatureFlag object so it's at the same place as the isEnabled logic? I don't feel strongly, just a thought.

I too don't feel strongly but we can address it in another PR if our opinions change in the future.

File: modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/bookmark/components/UpsellViewModel.kt (hunk 3 above)
Comment (ashiagr) [reply modules/features/player/src/main/java/au/com/shiftyjelly/pocketcasts/player/view/bookmark/components/UpsellViewModel.kt:33]: Thanks for sharing your thoughts! This was really tricky to test and each approach seemed to have its own pros and cons. At this time, I'd like to continue with the wrapper classes but I'll be on the lookout for a clearer testing approach that can be addressed in a future PR.

**Comment by ashiagr** [comment]
Thanks for a thorough review @mchowning! 

Regarding your comment,

> Obviously, we're still not handling the case where we want to have a feature in beta for multiple releases, but that still feels like an edge case we probably won't actually encounter.

I think it is still not clear to me how the feature should work for a production release prior to a Patron-exclusive release.  Should it not be exclusive to Patron in production until a full-access release? 

I'll go ahead and merge this PR since it works for current requirements but we can address it in another PR for any future possibility. 🙂
